### ISSUE TITLE
Bug with TypeDoesNotContainType (regression added in 4.3.0)

### ISSUE THREAD

**Issue by alexraputa** [vimeo/psalm#4797]
Hello,

for a code:

```php
<?php

use Psr\Container\ContainerInterface;
use Webmozart\Assert\Assert;

/** @psalm-return non-empty-string */
function __invoke(ContainerInterface $container): string
{
    $config = $container->get('config');

    Assert::isNonEmptyMap($config);
    Assert::keyExists($config, 'cookies');
    Assert::isNonEmptyMap($config['cookies']);
    Assert::keyExists($config['cookies'], 'domain');
    Assert::stringNotEmpty($config['cookies']['domain']);

    return $config['cookies']['domain'];
}
```

https://psalm.dev/r/929d0ab347

These errors are reported in 4.3.0:

```
ERROR: TypeDoesNotContainType - 15:13 - Cannot resolve types for $config['cookies']['domain'] - T does not contain non-empty-string
```

It was introduced in 4.3.0, was fine in 4.2.1.

But If I add redundant check `Assert::notEmpty` then there is no error.

https://psalm.dev/r/555b2972fb

**Comment by muglug** [comment]
Simplified to https://psalm.dev/r/dc37a77a75
